include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,394 @@
+//
+// Copyright (c) 2017 Damian Jarek (damian dot jarek93 at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/djarek/gabia
+//
+
+#ifndef GABIA_SECURE_SOCKET_HPP
+#define GABIA_SECURE_SOCKET_HPP
+
+#include <gabia/bytes.hpp>
+#include <gabia/crypto/aead.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/beast.hpp>
+#include <boost/endian/arithmetic.hpp>
+
+#define CORO_YIELD BOOST_ASIO_CORO_YIELD
+#define CORO_REENTER BOOST_ASIO_CORO_REENTER
+
+namespace gabia {
+
+namespace asio = boost::asio;
+namespace beast = boost::beast;
+
+template <typename DynamicBuffer>
+auto prepare_or_error(DynamicBuffer& buffer, size_t n,
+                      beast::error_code& error) {
+    boost::optional<typename DynamicBuffer::mutable_buffers_type> sequence;
+    try {
+        sequence = buffer.prepare(n);
+    } catch (const std::length_error&) {
+        error = asio::error::message_size;
+    }
+    return sequence;
+}
+
+template <typename Stream>
+class secure_socket {
+public:
+    using next_layer_type = typename std::remove_reference<Stream>::type;
+    using lowest_layer_type =
+        typename beast::get_lowest_layer<next_layer_type>::type;
+    constexpr static auto max_payload_length = 1024;
+
+    explicit secure_socket(asio::io_service& service)
+        : next_layer_stream{service} {}
+
+    template <class... Args>
+    secure_socket(crypto::aead_secrets secrets, Args&&... args)
+        : secrets{secrets}, next_layer_stream{std::forward<Args>(args)...} {}
+
+    auto& next_layer() { return next_layer_stream; }
+
+    const auto& next_layer() const { return next_layer_stream; }
+
+    auto& lowest_layer() { return next_layer().lowest_layer(); }
+
+    const auto& lowest_layer() const { return next_layer().lowest_layer(); }
+
+    auto& get_io_service() { return lowest_layer().get_io_service(); }
+
+    bool is_open() const { return lowest_layer().is_open(); }
+
+    void close() { lowest_layer().close(); }
+
+    void close(beast::error_code& error) { lowest_layer().close(error); }
+
+    template <typename ConstBufferSequence>
+    std::size_t write_some(const ConstBufferSequence& sequence,
+                           beast::error_code& error) {
+        error = {};
```

Comment:
The following code has the same effect and is faster (because it avoids the atomic fence for retrieving the default category)
```
error.assign(0, error.category());
```